### DISCUSSION TITLE
Training a model to forecast a TS for several subjects.

### DISCUSSION THREAD

**Discussion by darleybarreto** [nccr-itmo/FEDOT#431] [Q&A] [answered]
Hi folks, first I would like to say great job with FEDOT, it certainly is an incredible piece of software! For what I understood by looking at the examples, I can only train a model (or an automl pipeline + hyperopt) for each TS. So if there's several TS, one for each subject of the dataset, I would need to train one model for each one. Is that correct? Or there's a way of incrementally training, i.e., train for one subject, than the next, the next, and so on.

Thanks in advance!

**Comment by nicl-nno** [comment, marked as answer]
Hi, Darley.

Thank you for such positive feedback!

The described problem of incrementally training is similar to a hierarchical time series forecasting task. It is not implemented yet, but we already have several requests to add it. Conceptually, it fully satisfies the ideas behind FEDOT.

However, there is ambiguity in practical implementation (structure of input data, etc). It will be very useful if you can provide us the script with an example of the solution of described task using an approach that you noted as "need to train one model for each TS". 

It will help us to design a fully FEDOT-based solution to it in the future.

**Reply by darleybarreto** [reply]
Thank you for you quick response. I haven't coded what I said about a model for each TS, but my problem is similar to having a dataset of devices, e.g. electronic, to forecast some output such as throughput. Such dataset have several devices, where you have several indicators metrics, e.g. 5 (which are the features), across several timestamps, e.g., 100 days. In this case, a single device is basically a TS, so I would like to use FEDOT to predict the output, but using all devices. Is there a way of training a single model for all devices in the dataset?

**Reply by Dreamlone** [reply]
I prepared a picture so we can better understand each other :)

Do I understand correctly that you want to use option 1)? That is, to use time series **a** and **b** as predictors and pre-story of time series **c**, in order to predict series **c** 
![image](https://user-images.githubusercontent.com/52418437/134313934-5cde6c9c-85a0-42b5-a98c-59b400abb718.png)

Or do you want to predict all time series at once? In that case, I can suggest you use the iterative approach (shown in picture 2)). FEDOT can't predict time series for all time series at once, but it can do it sequentially

So, If option 1) is what you need, then I suggest you use the multivariate time series forecasting functionality in FEDOT. Check this notebooks ([forecast_vs_history.ipynb](https://github.com/nccr-itmo/FEDOT/blob/master/cases/industrial/forecast_vs_history.ipynb) and [multivariate_forecasting.ipynb](https://github.com/nccr-itmo/FEDOT/blob/master/cases/industrial/multivariate_forecasting.ipynb) in master branch). There the time series wind_power_kWh, diesel_time_h, wind_time_h, velocity_max_msec, velocity_mean_msec and tmp_grad are used to predict the time series diesel_fuel_kWh using AutoML.

**Reply by darleybarreto** [reply]
Wow @Dreamlone, thanks for the detailed response, really appreciate. It's basically the point `1)`. But having several instances, suppose I have a table of devices, where I have daily measurements of some performance metrics like this for every device (so several TS):

|device_id | measure_a | measure_b | measure_c|
|--- | --- | --- | ---|
| 001 | 12 | 0.23 | 36 |

Whereas my target is given in a monthly basis:
|device_id | month | target|
|--- | --- | ---|
| 001 | 12 | 12.25 |

If I had a single device, than `1)` would be perfect, but I have several, so I need to train the model to take those into account. In this particular case, would I need to run automl for each device?

**Reply by Dreamlone** [reply]
It looks like you have to run a separate model for each sensor if you want to predict multivariate time series (as in the picture above)

On the other hand, if you need to predict just one value for each device for every month, you can solve the regression problem. In this case, you can place devices in one table and train only one model

Such a table might look like this:
| month | device\_id | measure\_a\_day\_1 | measure\_a\_day\_2 | measure\_b\_day\_1 | measure\_b\_day\_2 | target |
| :---: | :--------: | :----------------: | :----------------: | :----------------: | :----------------: | :----: |
| 1     | 001        | ...                | ...                | ...                | ...                | ...    |
| 2     | 001        | ...                | ...                | ...                | ...                | ...    |
| 1     | 002        | ...                | ...                | ...                | ...                | ...    |
| 2     | 002        | ...                | ...                | ...                | ...                | ...    |

So for each month you can aggregate the daily values of your indicators **a**, **b** and **c**. 
Yes, there will be a lot of features columns in this case, but on the other hand FEDOT "knows" how to reduce the dimensions)

**Reply by darleybarreto** [reply]
In this case, wouldn't I lose information about trends & stuffs in the original TS?

**Reply by Dreamlone** [reply]
No, don't worry. Machine learning models will find the features that are most relevant to them and rely on those lags that contribute the most. That way, they can catch seasonality or low-frequency components of time series. On the other hand, you can do a discrete Fourier transform to get the frequency characteristics of the series, instead of lags. On such a table, you might be able to get better results

**Reply by darleybarreto** [reply]
I haven't worked with timeseries much, but If I have 12 months for each device, I have 12 rows on this table. Would the model understand which rows correspond to which devices and learn trends and seasonality?

**Reply by Dreamlone** [reply]
> Would the model understand which rows correspond to which devices and learn trends and seasonality?
Yes, if you place the row values opposite the value of the month and the device. 

Regarding seasonality and trending - just try running the model. If you do not get a satisfactory result, you can count the characteristics of the components of the time series, or get values after the Fourier transform. 

But first I advise you to try to build a model on your source features

**Reply by darleybarreto** [reply]
I will look into that. Thank you so much for your help!

**Reply by darleybarreto** [reply]
@Dreamlone I looked at this [classification](https://github.com/nccr-itmo/FEDOT/blob/9d5185b9bebb7a1de44cf2c2655daccb2c0ce459/examples/classification_with_tuning_example.py) example. If the data are in [this](https://github.com/nccr-itmo/FEDOT/discussions/431#discussioncomment-1368898) form, is automl a nice fit to find a pipeline? Or should I use the pipeline tuning?

**Reply by Dreamlone** [reply]
This pipeline tuning example shows how you can create your own pipeline without AutoML, and then tune it :) 

Yes, the example works for any tabular data and you can use it for [yours](https://github.com/nccr-itmo/FEDOT/discussions/431#discussioncomment-1368898) 

However, keep in mind that pipeline tuning is part of the AutoML algorithm. That is, when you run AutoML (as [here](https://github.com/ITMO-NSS-team/fedot-examples/blob/main/notebooks/latest/1_intro_to_automl.ipynb), you can check logs for phrase `Hyperparameters optimization start`) the model search goes through two steps: 
* composing (finding the optimal topology)
* tuning (tuning hyperparameters)

That is, you don't need to use tuning after applying AutoML. The found Pipeline has already gone through this procedure.

**Reply by darleybarreto** [reply]
Since automl-ing my train protocol takes lots of hours, what I had in mind was using automl to find a "decent" pipeline and then specify that pipeline by hand to train from scratch for more time, in this case also tuning it. What I mean by decent is making the `timeout` really small, so it probably wouldn't find the best one.
